File: packages/sdk-python/src/eidolondb/resources/entities.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .._client import EidolonDBClient
from .._types import Entity


class EntitiesResource:
    def __init__(self, client: EidolonDBClient):
        self._client = client

    def create(self, **kwargs: Any) -> Entity:
        return self._client.request("POST", "/entities", body=kwargs)

    def get(self, entity_id: str) -> Entity:
        return self._client.request("GET", f"/entities/{entity_id}")

    def list(
        self,
        *,
        type: Optional[str] = None,
        tag: Optional[str] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> List[Entity]:
        params: Dict[str, Any] = {
            "type": type,
            "tag": tag,
            "limit": limit,
            "offset": offset,
        }
        response = self._client.request("GET", "/entities", params=params)
        return response.get("entities", []) if isinstance(response, dict) else response

    def update(self, entity_id: str, **kwargs: Any) -> Entity:
        return self._client.request("PATCH", f"/entities/{entity_id}", body=kwargs)

    def delete(self, entity_id: str) -> None:
        self._client.request("DELETE", f"/entities/{entity_id}")

File: packages/sdk-js/src/resources/entities.ts
import { EidolonDBClient } from '../client.js';
import type { CreateEntityInput, Entity, ListEntitiesOptions, UpdateEntityInput } from '../types.js';

export class EntitiesResource {
  constructor(private readonly client: EidolonDBClient) {}

  /** Create an entity. */
  create(input: CreateEntityInput): Promise<Entity> {
    return this.client.request<Entity>('POST', '/entities', input);
  }

  /** Get an entity by ID. */
  get(id: string): Promise<Entity> {
    return this.client.request<Entity>('GET', `/entities/${id}`);
  }

  /** List entities. */
  async list(options?: ListEntitiesOptions): Promise<Entity[]> {
    const response = await this.client.request<{ entities: Entity[] }>('GET', '/entities', undefined, {
      query: options,
    });

    return response.entities;
  }

  /** Update an entity. */
  update(id: string, input: UpdateEntityInput): Promise<Entity> {
    return this.client.request<Entity>('PATCH', `/entities/${id}`, input);
  }

  /** Delete an entity. */
  async delete(id: string): Promise<void> {
    await this.client.request<{ deleted: boolean }>('DELETE', `/entities/${id}`);
  }
}
